src/lib.rs

```diff
@@ -317,9 +326,15 @@ unsafe extern "C" fn gmod13_open(state: *mut c_void) -> i32 {
     insert_function!(state, "InitializeTrain", |state| {
         let id = lua::Lcheckinteger(state, 1) as i32;
 
-        //let mut size = 0;
-        //let code = lua::Lchecklstring(state, 2, &mut size);
-
+        let mut size = 0;
+        let code_c_buf: *const c_schar = unsafe { lua::Lchecklstring(state, 2, &mut size) as *const i8};
+        let code_c_str: &CStr = unsafe { CStr::from_ptr(code_c_buf) };
+        let code_str_slice: &str = code_c_str.to_str().unwrap();
+        let code_str_buf: String = code_str_slice.to_owned();
```

Comment:
Lua source code can have zero byte in it. An it will be valid. So, there will be just one part of the buffer.
Also, as far as I can see, you are still passing old buffer size to the `train_thread`, so `luaL_loadbufferx` will try to acces unallocated memory. It might explode (bog-standard segfault), or might not (simple syntax error).
It will be better to construct slice from raw parts and then clone it to the Vector using `to_vec` method, but not `into_vec`. Also, much safer in this case.

Reply:
https://github.com/tam1192/gm_turbostroi_rust/commit/52714b6356fa9868dba1dc6892e5539b439a33f1

I tried with this commit and did not detect any anomalies in the operation.　
I wasn't familiar with RUST, so that was helpful.　Thanks.